src/activations.jl

```diff
@@ -237,65 +237,74 @@ for f in ACTIVATIONS
 end
 
 ## Faster, less accurate, versions of some.
-## TODO: mechanism to use automatically, and worry about GPUs.
+
+"""
+    tanh_fast(x::Float32)
+
+This is a faster but less accurate version of `tanh`.
+It has error of 5eps, instead of 1.5 eps for Julia's;
+under one less decimal digit. Usually about 10x faster.
+
+`tanh_new(x::Float16)` uses the same rational approximation,
+about twice the error of Julia's, about twice the speed
+(but still 5x slower than `Float32`). Might be better to remove this.
+"""
+@inline function tanh_fast(x::Float32)
+    x2 = abs2(x)
+    n = evalpoly(x2, (1.0f0, 0.1346604f0, 0.0035974074f0, 2.2332108f-5, 1.587199f-8))
+    d = evalpoly(x2, (1.0f0, 0.4679937f0, 0.026262015f0, 0.0003453992f0, 8.7767893f-7))
+    ifelse(x2 < 66f0, x * (n / d), sign(x))
+end
+
+# tanh_fast(x::Float16) # probably better. tanh_fast(x::Real) is very badly behaved for Float16.
+
+@inline function tanh_fast(x::Float16)
+    x2 = abs2(x)
+    n = evalpoly(x2, (Float16(1.0), Float16(0.1346), Float16(0.003597), Float16(2.235e-5)))
+    d = evalpoly(x2, (Float16(1.0), Float16(0.468), Float16(0.02626), Float16(0.0003455), Float16(9.0e-7)))
+    ifelse(x2 < Float16(18), x * (n / d), sign(x))
+end
 
 """
     tanh_fast(x) = exp(2x) / (1 + exp(2x))
 
 This is faster, but slightly less accurate, version of `tanh`.
+For `x::Float64`, this has errors up to about `3e-15`, sometimes `> 10 * eps(x)`,
+and should be about 2-5 times faster.
 
-For `x::Float64`, this has errors up to about `3e-15`, sometimes `> 10 * eps(x)`.
-For `x::Float32`, errors up to about `1f-6`. About 2-5 times faster.
+But `tanh_new(x::Float32)` is more important, and a different approximation.
 """
-@inline function tanh_fast(x)
+@inline function tanh_fast(x::Real)
    exp2x = @fastmath exp(x + x)
    y = (exp2x - 1) / (exp2x + 1)
```

Comment:
You should use expm1 here. It will be about 30% slower, but you will get correct answers for small x. This implimentation has catestrophic cancellation.

Reply:
Probably! Agree this looks dodgy, will re-check Float64 times & accuracy at some point.

I also wonder if there's a clever formula for `tanh(x)/x`, since it seems a little odd to get accuracy near 0 by expm1 when the desired result is close to linear?

Reply:
doesn't look like it.

Reply:
Switching to `expm1` seems to cost me the entire speed gain:
```
julia> xs = collect(0.01:0.001:10); ys = similar(xs);

julia> @btime @. $ys = tanh($xs);
  44.291 μs (0 allocations: 0 bytes)

julia> @btime @. $ys = tanh_fast($xs);  # with expm1
  47.958 μs (0 allocations: 0 bytes)

julia> @btime @. $ys = tanh_fast($xs);  # with exp(2x)-1
  11.500 μs (0 allocations: 0 bytes)
```
Accuracy, on this same range of points, improves 65eps -> 1.7eps relative worst case, but no change in absolute worst case 1.6e-16. So, as you say, the error is all for very small x.

Maybe it can evaluate some polynomial for `abs(x) < 0.13` or so, if 4eps is the goal.